### PR TITLE
[Feature] Queue all GCNotify requests

### LINKED ISSUE #10357: ✨ Queue all GCNotify messages
## ✨ Feature

Currently our API sends messages to GCNotify immediately "in-process".  We should switch these to be queued so that:
1. They can be retried if the request fails.
2. They can be delayed to stay within service rate limits.
3. Not delay the completion of web requests.
4. They won't cause the associated operation to fail in case of problems.

## 🕵️ Details

https://laravel.com/docs/10.x/queues

## ✅ Acceptance Criteria

- [ ] All API requests to GC Notify are queued and executed by worker processes.
- [ ] Failed jobs can be retried
- [ ] Service rate limits are respected

## 🛑 Blockers

```[tasklist]
### Blocked By
```

### PR DESCRIPTION
**Pull request by petertgiles**
🤖 Resolves #10357 

## 👋 Introduction

This branch updates the GCNotify notification channel to always queue every request instead of synchronously sending them.  It also adds a test notification and an artisan command for testing.

## 🧪 Testing

1. Rebuild the app and add the GCNotify env variables.  Use the team API key.
2. Shell into the web server container and start the queue worker `runuser -u www-data -- php /home/site/wwwroot/api/artisan queue:work`

### Test normal flow
- [ ] Try submitting a talent request with a whitelisted email address and observe that you receive an email.
- [ ] Try the test notification command `php artisan send-notifications:test (your email address) 1`
  - make sure you have a user set up with this email address

### Test errors
- [ ] Try sending a test notification to a non-whitelisted email address
  - Observe the job stays in the `jobs` table while it retires periodically
  - Observe the job eventually exceeds the exception limit and gets kicked over to the `failed_jobs` table

### Test the rate limiter
- [ ] Set the GCNOTIFY_MAX_REQUESTS_PER_MINUTE to just 2
- [ ] Try sending a bunch of valid messages
  -  `php artisan send-notifications:test (your email address) 6`
  - Observe the messages are received 2 per minute

### Azure
- deploy to dev and try the same tests
- no need to start the queue worker manually

## 📸 Screenshot

rate_limiter

## 🚚 Deployment

New variables (same for every vertical)
- GCNOTIFY_MAX_REQUESTS_PER_MINUTE=1000
- GCNOTIFY_TEMPLATE_TEST_EMAIL=ab0e7805-21db-4434-bd86-05f375c67304

### CODE REVIEW

File: api/app/Console/Commands/SendNotificationsTest.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        $emailAddress = $this->argument('emailAddress');
+        $numberOfMessages = $this->argument('number');
+
+        $successCount = 0;
+        $failureCount = 0;
+
+        $user = User::where('email', $emailAddress)->sole();
+
+        $notification = new Test('test', GcNotifyEmailChannel::class);
+        for ($i = 0; $i < $numberOfMessages; $i++) {
```
Comment (esizer) [inline api/app/Console/Commands/SendNotificationsTest.php:46]: One minor thing, I think this is causing one extra email to be sent :confused: I get one additional email than the number I pass in.

**Review by esizer — APPROVED** [review]
LGTM! :+1: 

I think there is one minor thing with the test command but other than that it works great! :tada:

File: api/app/Console/Commands/SendNotificationsTest.php (hunk 1 above)
Comment (petertgiles) [reply api/app/Console/Commands/SendNotificationsTest.php:46]: Sounds like it was just an old queued message.

File: api/app/Jobs/Middleware/GcNotifyRateLimited.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    /**
+     * Process the queued job.
+     *
+     * @param  \Closure(object): void  $next
+     */
+    public function handle(object $job, Closure $next): void
+    {
+        $executed = RateLimiter::attempt(
+            'gcnotify_api',
+            $perMinute = config('notify.client.max_requests_per_minute'),
```
Comment (mnigh) [inline api/app/Jobs/Middleware/GcNotifyRateLimited.php:19]: @petertgiles the [Laravel docs](https://laravel.com/docs/10.x/rate-limiting#basic-usage) use this in their example, but it seems very strange to define a variable that the linter flags as unused and when the type is an `int`. Am I a PHP 🦖 and am I missing something here?

## Screenshots

Screen Shot 2024-05-29 at 10 23 36
Screen Shot 2024-05-29 at 10 23 29

Comment (petertgiles) [reply api/app/Jobs/Middleware/GcNotifyRateLimited.php:19]: Yeah, Laravel sometimes does things in the documentation that the linter doesn't like.  I suppose they're still shoring up their type hinting.  Typed PHP is still relatively young I suppose.

Comment (petertgiles) [reply api/app/Jobs/Middleware/GcNotifyRateLimited.php:19]: You're welcome to try out experimenting with the GCNOTIFY_MAX_REQUESTS_PER_MINUTE variable to see how it works.

**Pull request merged**